rfcs/text/0008_service_status.md

```diff
@@ -0,0 +1,251 @@
+   */
+  plugins$: Observable<Record<string, ServiceStatus>>;
+
+  /**
+   * The default status of the plugin as specified in the "Status inheritance"
+   * section below.
+   * 
+   * Useful when overriding the defaults with `set`.
+   */
+  inherited$: Observable<ServiceStatus>;
+}
+```
+
+### HTTP API
+
+The HTTP endpoint should return basic information about the Kibana node as well as the overall system status and the status of each individual system.
+
+This API does not need to include UI-specific details like the existing API such as `uiColor` and `icon`.
+
+```ts
+/**
+ * Response type for the endpoint: GET /api/status
+ */
+interface StatusResponse {
+  /** server.name */
+  name: string;
+  /** server.uuid */
+  uuid: string;
+  /** Currently exposed by existing status API */
+  version: {
+    number: string;
+    build_hash: string;
+    build_number: number;
+    build_snapshot: boolean;
+  };
+  /** Similar format to existing API, but slightly different shape */
+  status: {
+    overall: ServiceStatus;
+    core: Record<string, ServiceStatus>;
+    plugins: Record<string, ServiceStatus>;
+  }
+}
+```
+
+## Behaviors
+
+### Levels
+
+Each member of the `ServiceStatusLevel` enum has specific behaviors associated with it:
+- **`available`**:
+  - All endpoints and apps associated with the service are accessible
+- **`degraded`**:
+  - All endpoints and apps are available by default
+  - Some APIs may return `503 Unavailable` responses. This must be implemented directly by the service.
+- **`unavailable`**:
+  - All endpoints return `503 Unavailable` responses by default. This is automatic.
+  - When trying to access any app associated with the unavailable service, the user is presented with an error UI with detail about the outage.
+- **`fatal`**:
+  - All endpoints (with some exceptions in Core) in Kibana return a `503 Unavailable` response by default. This is automatic.
+  - All applications redirect to the system-wide status page with detail about which services are down and any relevant detail.
+
+### Status inheritance
+
+By default, plugins inherit their status from all Core services and their dependencies on other plugins. The default status for a plugin follows this algorithm:
+- If any Core service is `degraded`, the plugin's status is `degraded`.
+- If any Core service is `unavailable`, the plugin's status is `unavailable`.
+- If any Core service is `fatal`, the plugin's status is `fatal`.
+- If any required dependency is `degraded`, the plugin's status is `degraded`.
+- If any required dependency is `unavailable`, the plugin's status is `unavailable`.
+- If any required dependency is `fatal`, the plugin's status is `fatal`.
+- If any optional dependency is `degraded`, the plugin's status is `degraded`.
+- If any optional dependency is `unavailable`, the plugin's status is `degraded`.
+- If any optional dependency is `fatal`, the plugin's status is `degraded`.
+- Otherwise, the plugin's status is `available`.
```

Comment:
Imho a little unclear regarding core/dependencies level prevalence/priority. What is the plugin's status if core is `degraded` but a required dependency is `unavailable`? If I follow the flow, it's `degraded` but I think it's actually `unavailable`, right?

Reply:
Good catch, I will make this more clear.

Reply:
Updated with more clear/correct pseudocode.